### PR TITLE
Provisioning: Dynamic user permissions

### LINKED ISSUE #1499: Provisioning of custom users/roles including global rights
- [ ] The feature was discussed internally in a TruBudget deployment project.
- [x] I have searched the [issues](https://github.com/openkfw/TruBudget/issues) of this repository and believe that this is not a duplicate.

## Description 💡

TruBudget's provisioning can be used to automatically generate TruBudget data, including users, groups and projects.
At the moment, the assignment of global permissions to users is hardcoded in [users_and_groups.js](https://github.com/openkfw/TruBudget/blob/main/provisioning/src/users_and_groups.js). 

A dynamic assignment of permissions defined in [users.json](https://github.com/openkfw/TruBudget/blob/main/provisioning/src/data/test/users.json) would allow us to create custom users with a custom selection of permissions. (For example, users with reading permissions only, another user with technical/blockchain permissions and a third one with project creation rights)

A possible user.json entry could look like:
```
{
    "id": "CustomProjectReader",
    "displayName": "Project Reader A",
    "password": "test",
    "permissions": ["project.list", "project.viewDetails"]
  }

```
As a developer, I would like to automatically provision TruBudget (functional) users including specific global permissions.

### PR DESCRIPTION
**Pull request by egdmrsy**
### Checklist

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/openkfw/TruBudget/blob/main/.github/CONTRIBUTING.md#open-a-pull-request).
- [ ] I fixed all necessary PR warnings
- [x] The commit history is clean
- [ ] The E2E tests are passing
- [x] If possible, the issue has been divided into more subtasks
- [x] I did a self review before requesting a review from another team member

### Description
Feature: adding dynamic permissions to users in provisioning
Closes #1499

### CODE REVIEW

File: provisioning/src/users_and_groups.js [hunk 1]
```diff
@@ -3,24 +3,49 @@ const {
...
-      log.info(`~> Adding user ${user.displayName}`);
-      await createUser(axios, user, organization);
-    }
-
-    await grantAllPermissionsToUser(axios, "mstein");
-    log.info("~> all Permissions granted to mstein");
 
-    await grantAllPermissionsToUser(axios, "jdoe");
-    log.info("~> all Permissions granted to jdoe");
+    await Promise.allSettled(users.map((user) => { log.info(`~> Adding user ${user.displayName}`); return createUser(axios, {id: user.id, displayName: user.displayName, password: user.password}, organization);}));
```
Comment (mayrmartin) [inline provisioning/src/users_and_groups.js:38]: what happens if the user is not created correctly? Maybe add some error handling?

File: provisioning/src/users_and_groups.js [hunk 2]
```diff
@@ -3,24 +3,49 @@ const {
...
+    const usersWithPermissions = users.filter((user) => Array.isArray(user.permissions) && user.permissions.length > 0);
+    for (const user of usersWithPermissions) {
+      if (user.permissions.includes("all")) {
+        log.info(`~> all Permissions granted to ${user.id}`);
+        await grantAllPermissionsToUser(axios, user.id);
```
Comment (mayrmartin) [inline provisioning/src/users_and_groups.js:43]: what happens if the permission can not be granted? Maybe add some error handling?

File: provisioning/src/users_and_groups.js [hunk 3]
```diff
@@ -3,24 +3,49 @@ const {
...
+      } 
+      else {
+        await Promise.allSettled(user.permissions.filter(intent => allowedIntents.includes(intent)).map((intent) => { log.info(`~> Granting permission ${intent} to ${user.id}`); return grantGlobalPermissionToUser(axios, intent, user.id);}));
```
Comment (mayrmartin) [inline provisioning/src/users_and_groups.js:46]: what happens if the permission can not be granted? Maybe add some error handling?

File: provisioning/src/users_and_groups.js [hunk 4]
```diff
@@ -34,23 +59,12 @@ const provisionUsers = async (axios, folder, organization) => {
...
   try {
-    const users = readJsonFile(folder + "users.json");
     const groups = readJsonFile(folder + "groups.json");
-    for (const group of groups) {
-      log.info(`~> Adding group ${group.displayName}`);
-      await createGroup(axios, group);
+    await Promise.allSettled(groups.map((group) => createGroup(axios, group)));
+    const groupsWithUsers = groups.filter((group) => Array.isArray(group.users) && group.users.length > 0);
+    for (const group of groupsWithUsers) {
+      await Promise.allSettled(group.users.map((user) => addUserToGroup(axios, group.id, user)));
```
Comment (mayrmartin) [inline provisioning/src/users_and_groups.js:66]: what happens if the user can not be added? Maybe add some error handling?

File: provisioning/src/users_and_groups.js [hunk 5]
```diff
@@ -34,23 +59,12 @@ const provisionUsers = async (axios, folder, organization) => {
 
 const provisionGroups = async (axios, folder) => {
   try {
-    const users = readJsonFile(folder + "users.json");
     const groups = readJsonFile(folder + "groups.json");
-    for (const group of groups) {
-      log.info(`~> Adding group ${group.displayName}`);
-      await createGroup(axios, group);
+    await Promise.allSettled(groups.map((group) => createGroup(axios, group)));
```
Comment (mayrmartin) [inline provisioning/src/users_and_groups.js:63]: what happens if the group can not be created? Maybe add some error handling?

**Review by mayrmartin — CHANGES_REQUESTED** [review]

**Review by mayrmartin — APPROVED** [review]

File: provisioning/src/users_and_groups.js [hunk 6]
```diff
@@ -3,24 +3,26 @@ const {
...
-      await createUser(axios, user, organization);
-    }
-
-    await grantAllPermissionsToUser(axios, "mstein");
-    log.info("~> all Permissions granted to mstein");
 
-    await grantAllPermissionsToUser(axios, "jdoe");
-    log.info("~> all Permissions granted to jdoe");
+    await Promise.all(users.map((user) => { log.info(`~> Adding user ${user.displayName}`); return createUser(axios, {id: user.id, displayName: user.displayName, password: user.password}, organization);}));
+    const usersWithPermissions = users.filter((user) => Array.isArray(user.permissions) && user.permissions.length > 0);
```
Comment (georgimld) [inline provisioning/src/users_and_groups.js:16]: when would `Array.isArray(user.permissions)` be false?

**Review by georgimld — APPROVED** [review]

**Pull request merged**